src/main/java/io/github/thebusybiscuit/slimefun4/core/networks/cargo/ItemFilter.java

```diff
@@ -0,0 +1,175 @@
+     *            The {@link Block}
+     */
+    public ItemFilter(@Nonnull Block b) {
+        update(b);
+    }
+
+    /**
+     * This updates or refreshes the {@link ItemFilter} to copy the settings
+     * from the given {@link Block}. It takes a new snapshot.
+     * 
+     * @param b
+     *            The {@link Block}
+     */
+    public void update(@Nonnull Block b) {
+        // Store the returned Config instance to avoid heavy calls
+        Config blockData = BlockStorage.getLocationInfo(b.getLocation());
+        String id = blockData.getString("id");
+        SlimefunItem item = SlimefunItem.getByID(id);
+        BlockMenu menu = BlockStorage.getInventory(b.getLocation());
+
+        if (item == null || menu == null) {
+            // Don't filter for a non-existing item (safety check)
+            clear(false);
+        } else if (id.equals("CARGO_NODE_OUTPUT")) {
+            // Output Nodes have no filter, allow everything
+            clear(true);
+        } else {
+            this.items.clear();
+            this.checkLore = Objects.equals(blockData.getString("filter-lore"), "true");
+            this.defaultValue = !Objects.equals(blockData.getString("filter-type"), "whitelist");
+
+            for (int slot : CargoUtils.FILTER_SLOTS) {
+                ItemStack stack = menu.getItemInSlot(slot);
+
+                if (stack != null && stack.getType() != Material.AIR) {
+                    this.items.add(new ItemStackWrapper(stack));
+                }
+            }
+        }
+
+        this.dirty = false;
+    }
+
+    /**
+     * This will clear the {@link ItemFilter} and reject <strong>any</strong>
+     * {@link ItemStack}.
+     * 
+     * @param defaultValue
+     *            The new default value.
+     */
+    private void clear(boolean defaultValue) {
+        this.items.clear();
+        this.checkLore = false;
+        this.defaultValue = defaultValue;
+    }
+
+    /**
+     * Whether this {@link ItemFilter} is outdated and needs to be refreshed.
+     * 
+     * @return Whether the filter is outdated.
+     */
+    public boolean isDirty() {
+        return dirty;
+    }
+
+    /**
+     * This marks this {@link ItemFilter} as dirty / outdated.
+     */
+    public void markDirty() {
+        this.dirty = true;
+    }
+
+    @Override
+    public boolean test(@Nonnull ItemStack item) {
+        // An empty Filter does not need to be iterated over.
+        if (items.isEmpty()) {
+            return defaultValue;
+        }
+
+        int potentialMatches = 0;
+
+        // This is a first check for materials to see if we might even have any match.
+        // If there is no potential match then we won't need to perform the quite
+        // intense operation .getItemMeta()
+        for (ItemStackWrapper stack : items) {
+            if (stack.getType() == item.getType()) {
+                // We found a potential match based on the Material
+                potentialMatches++;
+            }
+        }
```

Comment:
Not sure I'm the biggest fan of having another loop just for the sake of saving 1 #getItemMeta call on `item`. Especially since both the loops are for an Array meaning O(n). If it was a hash table and be O(1) then sure maybe but yeah. I feel like we're losing performance to try and improve performance.

Reply:
getItemMeta() is a very heavy function that can explode quite quickly.
This is a bit ugly but it does improve performance overall actually.
The horror of ItemMeta cloning is really worth this sacrifice here sadly.

Reply:
Also, keep in mind that `items` has a maximum capacity of 9. So the impact will be quite small.

Reply:
(In the meantime I approved since I assume we're just waiting on Walshy to decide he believes you, take this as a bump for the blob)